### PR TITLE
Allow passing PullConsumer settings as keyword list

### LINKED ISSUE #20: Allow passing PullConsumer settings as keyword list along with map
We currently expect that `PullConsumer` settings is a map with specified keys. From end-user perspective, it would be more pleasant if we would also accept keyword lists here (with same keys of course), because Elixir has special syntax for keyword lists at the end of some expressions:

```elixir
children = [
  {ExamplePullConsumer,
    %{
      connection_name: :gnat,
      stream_name: stream_name,
      consumer_name: consumer_name,
    }
  }
]
```

would be nicer if it could be written as

```elixir
children = [
  {ExamplePullConsumer,
    connection_name: :gnat,
    stream_name: stream_name,
    consumer_name: consumer_name,
  }
]
```

### PR DESCRIPTION
**Pull request by sswrk**
close #20

### CODE REVIEW

File: lib/jetstream/pull_consumer.ex [hunk 1]
```diff
@@ -71,13 +69,21 @@ defmodule Jetstream.PullConsumer do
...
+            optional(:connection_retry_timeout) => pos_integer(),
+            optional(:connection_retries) => pos_integer()
+          }
+          | [
+              connection_name: GenServer.server(),
+              stream_name: binary(),
+              consumer_name: binary(),
+              connection_retry_timeout: pos_integer(),
+              connection_retries: pos_integer()
+            ]
```
Comment (sswrk) [inline lib/jetstream/pull_consumer.ex:86]: Correct me if I'm wrong, but I think there's no way to emphasize that some Keyword list keys are required and others are optional.

I wonder which version is more correct - the current one or

```elixir
@type option ::
          {:connection_name, GenServer.server()}
          | {:stream_name, binary()}
          | {:consumer_name, binary()}
          | {:connection_retry_timeout, pos_integer()}
          | {:connection_retries, pos_integer()}

  @type settings ::
          %{
            :connection_name => GenServer.server(),
            :stream_name => binary(),
            :consumer_name => binary(),
            optional(:connection_retry_timeout) => pos_integer(),
            optional(:connection_retries) => pos_integer()
          }
          | [option()]
```

Comment (mkaput) [reply lib/jetstream/pull_consumer.ex:86]: Usually what I observed is that type specs are unspecified (I.e. `Keyword.t() | map()`), while parameters are only documented as prose in `@moduledoc`

File: lib/jetstream/pull_consumer.ex [hunk 2]
```diff
@@ -105,14 +111,17 @@ defmodule Jetstream.PullConsumer do
   @spec start_link(module :: module(), settings :: settings(), options :: GenServer.options()) ::
           GenServer.on_start()
   def start_link(module, settings, options \\ []) do
-    settings = %{
-      settings: settings,
+    state = %{
+      settings: settings_to_map(settings),
```
Comment (mkaput) [inline lib/jetstream/pull_consumer.ex:115]:
```suggestion
      settings: Map.new(settings),
```

File: lib/jetstream/pull_consumer.ex [hunk 3]
```diff
@@ -45,7 +45,20 @@ defmodule Jetstream.PullConsumer do
...
+  * `:consumer_name` - name of an existing consumer pointing at the stream.
+
+  You can also pass the optional ones:
+
+  * `:connection_retry_timeout` - a duration in milliseconds after which the PullConsumer which failed to
+    establish NATS connection retries. Defaults to `1_000`.
+  * `:connection_retreis` - a number of attempts the PullConsumer will make to establish the NATS connection.
+    When this value is exceeded, the PullConsumer stops with the `:timeout` reason.
+
+  The settings can be passed both as a map and as a keyword list.
```
Comment (mkaput) [inline lib/jetstream/pull_consumer.ex:61]:
```suggestion
  The settings can be passed as either a map or a keyword list.
```

File: lib/jetstream/pull_consumer.ex [hunk 4]
```diff
@@ -45,7 +45,20 @@ defmodule Jetstream.PullConsumer do
   end
   ```
 
-  Note that the stream and its consumer, which names were given, must exist.
+  The following settings must be provided:
+
+  * `:connection_name` - Gnat connection or `Gnat.ConnectionSupervisor` name/PID
```
Comment (mkaput) [inline lib/jetstream/pull_consumer.ex:50]:
```suggestion
  * `:connection_name` - Gnat connection or `Gnat.ConnectionSupervisor` name/PID.
```

File: lib/jetstream/pull_consumer.ex [hunk 5]
```diff
@@ -45,7 +45,20 @@ defmodule Jetstream.PullConsumer do
...
+  * `:stream_name` - name of an existing string the consumer will consume messages from.
+  * `:consumer_name` - name of an existing consumer pointing at the stream.
+
+  You can also pass the optional ones:
+
+  * `:connection_retry_timeout` - a duration in milliseconds after which the PullConsumer which failed to
+    establish NATS connection retries. Defaults to `1_000`.
+  * `:connection_retreis` - a number of attempts the PullConsumer will make to establish the NATS connection.
```
Comment (mkaput) [inline lib/jetstream/pull_consumer.ex:58]:
```suggestion
  * `:connection_retries` - a number of attempts pull consumer will make to establish the NATS connection.
```

**Review by mkaput — CHANGES_REQUESTED** [review]

**Review by mkaput — APPROVED** [review]

**Pull request merged**